### PR TITLE
Prepare auto-updating installer with default plugins and customizations

### PR DESCRIPTION
**Pull request by jfaltermeier**
#### What it does

* add preinstalled theiaPlugins to package-json of electron-app (preparation for #5)
* add lerna to enable a multi-package repository for electron app + customization packages
* add a custom main js that sets the environment variable for preinstalled plugins
* add a customization module for the installer at theia-example-updater. This may be used to enable updates (#13), to customize the about dialog (#10), and to customize the getting started page (#9)

Signed-off-by: Johannes Faltermeier <jfaltermeier@eclipsesource.com>
Signed-off-by: Lucas Koehler <lkoehler@eclipsesource.com>
Contributed on behalf of STMicroelectronics

#### How to test

Run `yarn package` and check if the package is still working. 

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

**Comment by koegel** [comment]
@thegecko would you be available for reviewing this PR?

File: theia-example-updater/package.json [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+  },
+  "homepage": "https://github.com/eclipse-theia/theia-example",
+  "files": [
+    "lib",
+    "src"
+  ],
+  "scripts": {
+    "prepare": "yarn run clean && yarn run build",
+    "clean": "rimraf lib",
+    "build": "tsc --build base.tsconfig.json"
```
Comment (thegecko) [inline theia-example-updater/package.json:42]: This should use a local `tsconfig.json` file which inherits from the base one. The base one should be put in the root of the monorepo for use/exension by other packages

**Review by thegecko — CHANGES_REQUESTED** [review]
I need to look into getting PRs built on jenkins so we know a merge won't break master. Aside from that there are a few minor comments

File: electron-app/electron-builder.yml [hunk 2]
```diff
@@ -1,17 +1,27 @@
-appId: eclipse.theia
+appId: eclipse.theia.java
```
[the remark this answers is not in the document]
Comment (jfaltermeier) [reply electron-app/electron-builder.yml]: This was an oversight and I'll remove it. Initially the only plugins that were preinstalled during testing were Java related.

File: electron-app/electron-builder.yml [hunk 3]
```diff
@@ -50,3 +60,7 @@ deb:
   artifactName: ${productName}-${version}.${ext}
 
 afterPack: ./scripts/after-pack.js
+
+publish:
```
[the remark this answers is not in the document]
Comment (jfaltermeier) [reply electron-app/electron-builder.yml]: I'll remove it. We can add the proper configuration with #13 for deployment on a generic HTTP server like the Eclipse servers.

**Comment by jfaltermeier** [comment]
Thanks for the review. We addressed the comments and rebased the changes on the latest master

**Comment by thegecko** [comment]
Thanks for the updates. As I find time I'm modifying jenkins to build PRs and see whether this passes.

**Comment by thegecko** [comment]
Can you rebase this on master and resolve the conflict?

**Comment by jfaltermeier** [comment]
Sure, conflict is resolved

**Comment by thegecko** [comment]
[building...](https://ci.eclipse.org/theia/job/theia-example/view/change-requests/job/PR-15/)

**Comment by thegecko** [comment]
The windows build failed with:

```
cannot find specified resource "LICENSE", nor relative to "C:\Users\genie.theia\jenkins_agent\workspace\theia-example_PR-15\electron-app\resources", neither relative to project dir ("C:\Users\genie.theia\jenkins_agent\workspace\theia-example_PR-15\electron-app")
```

Which suggests a file path is wrong with `LICENSE`.

There's also Mac issues, but that's likely to do with the `node` upgrade being done on there.

**Comment by thegecko** [comment]
Thanks, looks like the changes have fixed Windows. We are just waiting on the node upgrade on the mac agent:

https://bugs.eclipse.org/bugs/show_bug.cgi?id=568685

**Comment by thegecko** [comment]
The node upgrade is complete and this PR now fails during signing.

This is due to how jenkins undertakes builds. As this this PR is untrusted (opened from a fork), therefore it uses the origin `jenkinsfile` (from `master`) to undertake the build and due to the output files moving from `dist` it can't find them.

There are a few ways forward:
- Slacken the security to trust all PRs from forks (not ideal)
- Split the `Jenkinsfile` update out of this PR and merge it separately from a branch (will break `master`)
- Have this PR submitted from a branch rather than a fork

Are you able to reopen this from a branch (if you have the rights?)

**Comment by thegecko** [comment]
Apologies, I've just merged the 1.8.0 release ahead of this too so another rebase is required.

**Comment by jfaltermeier** [comment]
I've updated the PR to 1.8. Also I've tried to push my changes to a branch in this repository, however I don't have the rights. 
So should I remove the jenkinsfile changes from this PR? (I think we've only updated paths, because the dist directory is now nested within the electron-app directory rather than at the root)

**Comment by thegecko** [comment]
> I've updated the PR to 1.8

Thanks!

> should I remove the jenkinsfile changes from this PR?

I guess it won't work without that change unless you can think of a way of it can keep the paths?

> I've tried to push my changes to a branch in this repository, however I don't have the rights

OK, I'll see what I can do

**Comment by thegecko** [comment]
@jfaltermeier can you rebase this PR again and use the [`distFolder` variable](https://github.com/eclipse-theia/theia-example/blob/master/Jenkinsfile#L6) in the `jenkinsfile` rather than the inline changes to `dist`?

I can then run your PR end-to-end.

**Comment by jfaltermeier** [comment]
Thanks, jenkinsfile is updated now

**Comment by thegecko** [comment]
Looks like it built! Merging.

**Comment by thegecko** [comment]
Merge succeeded, the latest 1.8 release here should include your changes, but I haven't tested it:

https://download.eclipse.org/theia/

**Comment by jfaltermeier** [comment]
Thank you, looks good!

**Pull request merged**